src/bms/parse/model.rs

```diff
@@ -0,0 +1,1227 @@
+            }
+        }
+        Ok(())
+    }
+}
+
+impl Bms {
+    /// Gets the time of last any object including visible, BGM, BPM change, section length change and so on.
+    ///
+    /// You can't use this to find the length of music. Because this doesn't consider that the length of sound.
+    pub fn last_obj_time(&self) -> Option<ObjTime> {
+        let obj_last = self
+            .notes
+            .objs
+            .values()
+            .flatten()
+            .map(Reverse)
+            .sorted()
+            .next()
+            .map(|Reverse(obj)| obj.offset);
+        let bpm_last = self
+            .arrangers
+            .bpm_changes
+            .last_key_value()
+            .map(|(&time, _)| time);
+        let section_len_last =
+            self.arrangers
+                .section_len_changes
+                .last_key_value()
+                .map(|(&time, _)| ObjTime {
+                    track: time,
+                    numerator: 0,
+                    denominator: 4,
+                });
+        let stop_last = self.arrangers.stops.last_key_value().map(|(&time, _)| time);
+        let bga_last = self
+            .graphics
+            .bga_changes
+            .last_key_value()
+            .map(|(&time, _)| time);
+        [obj_last, bpm_last, section_len_last, stop_last, bga_last]
+            .into_iter()
+            .max()
+            .flatten()
+    }
+
+    /// Calculates a required resolution to convert the notes time into pulses, which split one quarter note evenly.
+    pub fn resolution_for_pulses(&self) -> u64 {
+        use num::Integer;
+
+        let mut hyp_resolution = 1u64;
+        for obj in self.notes.objs.values().flatten() {
+            hyp_resolution = hyp_resolution.lcm(&obj.offset.denominator);
+        }
+        for bpm_change in self.arrangers.bpm_changes.values() {
+            hyp_resolution = hyp_resolution.lcm(&bpm_change.time.denominator);
+        }
+        hyp_resolution
+    }
+}
+
+impl Arrangers {
+    /// Adds a new BPM change object to the notes.
+    pub fn push_bpm_change(
+        &mut self,
+        bpm_change: BpmChangeObj,
+        prompt_handler: &mut impl PromptHandler,
+    ) -> Result<()> {
+        if let Some(existing) = self.bpm_changes.insert(bpm_change.time, bpm_change.clone()) {
+            match prompt_handler.handle_duplication(PromptingDuplication::BpmChangeEvent {
+                time: bpm_change.time,
+                older: &existing,
+                newer: &bpm_change,
+            }) {
+                DuplicationWorkaround::UseOlder => {
+                    self.bpm_changes.insert(bpm_change.time, existing);
+                }
+                DuplicationWorkaround::UseNewer => {
+                    // Already inserted the newer one
+                }
+                DuplicationWorkaround::Warn => {
+                    return Err(ParseWarning::PromptHandlerWarning);
+                }
+            }
+        }
```

Comment:
This may occur two insertions. How about checking whether it was contained with BTreeMap's `contains_key` method?